Lib/inspect.py

```diff
@@ -801,25 +801,48 @@ def findsource(object):
         return lines, 0
 
     if isclass(object):
-        name = object.__name__
-        pat = re.compile(r'^(\s*)class\s*' + name + r'\b')
-        # make some effort to find the best matching class definition:
-        # use the one with the least indentation, which is the one
-        # that's most probably not inside a function definition.
-        candidates = []
-        for i in range(len(lines)):
-            match = pat.match(lines[i])
-            if match:
-                # if it's at toplevel, it's already the best one
-                if lines[i][0] == 'c':
-                    return lines, i
-                # else add whitespace to candidate list
-                candidates.append((match.group(1), i))
-        if candidates:
-            # this will sort by whitespace, and by line number,
-            # less whitespace first
-            candidates.sort()
-            return lines, candidates[0][1]
+        # Lazy import ast because it's relatively heavy and
+        # it's not used for other than this part.
+        import ast
+
+        class ClassFinder(ast.NodeVisitor):
+
+            def visit_FunctionDef(self, node):
+                stack.append(node.name)
+                stack.append('<locals>')
+                self.generic_visit(node)
+                stack.pop()
+                stack.pop()
+
+            visit_AsyncFunctionDef = visit_FunctionDef
+
+            def visit_ClassDef(self, node):
+                nonlocal line_number
+                stack.append(node.name)
+                if qualname == '.'.join(stack):
+                    # Return the decorator for the class if present
+                    if node.decorator_list:
+                        line_number = node.decorator_list[0].lineno
+                    else:
+                        line_number = node.lineno
+
+                    # decrement by one since lines starts with indexing by zero
+                    line_number -= 1
+                    raise StopIteration(line_number)
```

Comment:
I don't like this way of interrupting the AST traversal.  Just declare a custom exception, raise it, and then catch it.